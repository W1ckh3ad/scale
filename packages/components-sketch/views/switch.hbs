<div data-sketch-artboard="Switch" class="artboard">
    <h1 class="title--artboard">Switch</h1>

    <h2 class="title--artboard-section">Selected On</h2>
    <div class="agent-states--grid">
        <div class="agent-states--item">
            <p class="agent-states--label">:available</p>
            <div class="switch-wrapper" data-sketch-key="switch-1">Switch Label<scale-switch text-on="I" text-off="0" active data-sketch-key="internal-switch-1"></scale-switch></div>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <div class="switch-wrapper" data-sketch-key="switch-2">Switch Label<scale-switch text-on="I" text-off="0" active data-sketch-state="hover" data-sketch-key="internal-switch-2"></scale-switch></div>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <div class="switch-wrapper" data-sketch-key="switch-3">Switch Label<scale-switch text-on="I" text-off="0" active data-sketch-state="focus" data-sketch-key="internal-switch-3"></scale-switch></div>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:pressed</p>
            <div class="switch-wrapper" data-sketch-key="switch-4">Switch Label<scale-switch text-on="I" text-off="0" active data-sketch-state="active" data-sketch-key="internal-switch-4"></scale-switch></div>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <div class="switch-wrapper" data-sketch-key="switch-5">Switch Label<scale-switch text-on="I" text-off="0" active disabled data-sketch-key="internal-switch-5"></scale-switch></div>
        </div>
    </div>

    <h2 class="title--artboard-section">Selected Off</h2>
    <div class="agent-states--grid">
        <div class="agent-states--item">
            <p class="agent-states--label">:available</p>
            <div class="switch-wrapper" data-sketch-key="switch-6">Switch Label<scale-switch data-sketch-key="internal-switch-10"></scale-switch></div>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <div class="switch-wrapper" data-sketch-key="switch-7">Switch Label<scale-switch text-on="I" text-off="0" data-sketch-state="hover" data-sketch-key="internal-switch-6"></scale-switch></div>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <div class="switch-wrapper" data-sketch-key="switch-8">Switch Label<scale-switch text-on="I" text-off="0" data-sketch-state="focus" data-sketch-key="internal-switch-7"></scale-switch></div>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:pressed</p>
            <div class="switch-wrapper" data-sketch-key="switch-9">Switch Label<scale-switch text-on="I" text-off="0" data-sketch-state="active" data-sketch-key="internal-switch-8"></scale-switch></div>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <div class="switch-wrapper" data-sketch-key="switch-10">Switch Label<scale-switch text-on="I" text-off="0" disabled data-sketch-key="internal-switch-9"></scale-switch></div>
        </div>
    </div>
</div>


<script src="/symbol_names.js"></script>
<script>
    Array.from(document.querySelectorAll(".switch-wrapper")).forEach(btn => {
        btn.dataset.sketchSymbol = `Switch / ${
                order(1, btn.firstElementChild.hasAttribute('active') ? 'Selected On' : 'Selected Off')
            } / ${
                order(4, getStateName(btn.firstElementChild))
            }`;
    });
</script>