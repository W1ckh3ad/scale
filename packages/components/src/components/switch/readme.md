# scale-switch



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                        | Type      | Default     |
| ------------- | -------------- | ---------------------------------- | --------- | ----------- |
| `active`      | `active`       | (optional) Active switch           | `boolean` | `false`     |
| `customClass` | `custom-class` | (optional) Switch class            | `string`  | `''`        |
| `disabled`    | `disabled`     | (optional) Disabled switch         | `boolean` | `false`     |
| `iconOff`     | `icon-off`     | (optional) icon when switch is off | `string`  | `undefined` |
| `iconOn`      | `icon-on`      | (optional) icon when switch is on  | `string`  | `undefined` |
| `iconSize`    | `icon-size`    | (optional)  Icon size              | `number`  | `16`        |
| `styles`      | `styles`       | (optional) Injected jss styles     | `any`     | `undefined` |
| `textOff`     | `text-off`     | (optional) list Icon size          | `string`  | `undefined` |
| `textOn`      | `text-on`      | (optional) List item icon          | `string`  | `undefined` |


## Dependencies

### Depends on

- [scale-icon](../icon)

### Graph
```mermaid
graph TD;
  scale-switch --> scale-icon
  style scale-switch fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
