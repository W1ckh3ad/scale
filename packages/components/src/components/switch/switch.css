.switch {
  --switch-size: 20px;
  --switch-offset: 2px;
  --switch-color-off: #262626;
  --switch-color-on: var(--scl-color-primary);
  --switch-color-active: var(--scl-color-primary-active);
  --switch-color-hover: var(--scl-color-primary-hover);
  --switch-color-disabled: var(--scl-color-background-disabled);
  --switch-color-state: var(--switch-color-off);
  --switch-color-state-contrast: var(--scl-color-background-default);
  --switch-color-background: var(--switch-color-state-contrast);
  --switch-color-toggle: var(--switch-color-state);
  --switch-color-text: var(--switch-color-state);

  display: inline-block;
  position: relative;
}

.switch label {
  cursor: pointer;
}

.switch input {
  width: 0;
  height: 0;
  opacity: 0;
  position: absolute;
}

.switch .switch__wrapper {
  position: relative;
  border: 1px solid var(--switch-color-state);
  border-radius: calc(var(--switch-size) / 2);
  width: calc(2 * var(--switch-size) - var(--switch-offset));
  height: calc(var(--switch-size) - var(--switch-offset));
  display: inline-flex;
  justify-content: space-between;
  background: var(--switch-color-background);
  transition-property: background;
  transition-duration: var(--scl-motion-duration-fast);
  transition-timing-function: var(--scl-motion-easing-default);
}

.switch .switch__text {
  display: flex;
  width: 100%;
  color: var(--switch-color-text);
  justify-content: space-evenly;
  font-size: var(--scl-font-size-12);
  line-height: calc(var(--switch-size) - var(--switch-offset));
  transition-property: color;
  transition-duration: var(--scl-motion-duration-fast);
  transition-timing-function: var(--scl-motion-easing-default);
}

.switch .switch__text:before {
  content: 'I';
}

.switch .switch__text:after {
  content: '0';
  position: relative;
  right: -1px;
}

.switch .switch__toggle {
  position: absolute;
  border-radius: 50%;
  padding: calc(0.5 * var(--switch-size) - var(--switch-offset));
  margin: calc(-0.5 * var(--switch-size) + var(--switch-offset));
  background: var(--switch-color-toggle);
  left: calc(0.5 * var(--switch-size) - 0.5 * var(--switch-offset));
  top: 50%;
  transition-property: transform, background;
  transition-duration: var(--scl-motion-duration-fast);
  transition-timing-function: var(--scl-motion-easing-default);
}

.switch input:checked:disabled + .switch__wrapper .switch__toggle,
.switch
  input:not(:checked):not(:disabled):active
  + .switch__wrapper
  .switch__toggle,
.switch
  input:checked:not(:disabled):not(:active)
  + .switch__wrapper
  .switch__toggle {
  transform: translateX(var(--switch-size));
}

.switch .switch__label {
  cursor: inherit;
  color: var(--switch-color-text);
  margin-left: 1rem;
  transition-property: color;
  transition-duration: var(--scl-motion-duration-fast);
  transition-timing-function: var(--scl-motion-easing-default);
}

.switch--focus-visible-not-supported input:focus ~ .switch__wrapper {
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}

.switch--focus-visible-supported input:focus-visible ~ .switch__wrapper {
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}

.switch input:checked ~ .switch__wrapper {
  --switch-color-state: var(--switch-color-on);
  --switch-color-background: var(--switch-color-state);
  --switch-color-toggle: var(--switch-color-state-contrast);
  --switch-color-text: var(--switch-color-state-contrast);
}

/* Needs to placed after :checked but before :active */
.switch:hover,
.switch:hover input ~ .switch__wrapper {
  --switch-color-state: var(--switch-color-hover);
}

.switch input:not(:disabled):active ~ .switch__wrapper {
  --switch-color-state: var(--switch-color-active);
  --switch-color-background: var(--switch-color-state);
  --switch-color-toggle: var(--switch-color-state-contrast);
  --switch-color-text: var(--switch-color-state-contrast);
}

.switch input:checked:not(:disabled):active ~ .switch__wrapper {
  --switch-color-state: var(--switch-color-active);
  --switch-color-background: var(--switch-color-state-contrast);
  --switch-color-toggle: var(--switch-color-active);
  --switch-color-text: var(--switch-color-active);
}

.switch input:not(:disabled):active ~ .switch__label {
  --switch-color-text: var(--switch-color-active);
}

/* disabled states */

.switch.switch--disabled {
  --switch-color-state: var(--switch-color-disabled);
  --switch-color-hover: var(--switch-color-disabled);
  --switch-color-toggle: var(--switch-color-disabled);
  --switch-color-text: var(--switch-color-disabled);
}

.switch.switch--disabled label {
  cursor: not-allowed;
}

.switch.switch--disabled input:checked ~ .switch__wrapper {
  --switch-color-state: var(--switch-color-disabled);
  --switch-color-toggle: var(--switch-color-state-contrast);
  --switch-color-text: var(--switch-color-state-contrast);
}
