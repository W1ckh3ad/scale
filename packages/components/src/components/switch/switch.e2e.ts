import { newE2EPage } from '@stencil/core/testing';

describe('scale-switch', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-switch></scale-switch>');
    const element = await page.find('scale-switch');
    expect(element).toHaveClass('hydrated');
  });

  // it('should not toggle to active when switch is disabled ', async () => {
  //   const page = await newE2EPage({
  //     html: `<t-switch disabled="true" active="false"></t-switch>`,
  //   });
  //   const ele = await page.find('t-switch >>> .switch');
  //   await ele.click();
  //   expect(ele).not.toContain('.switch--active');
  // });
});
