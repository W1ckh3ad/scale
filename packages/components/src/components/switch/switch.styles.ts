import { JssStyle } from 'jss';

export const styles: JssStyle = {
  switch: {
    margin: 0,
    display: 'inline-block',
    position: 'relative',
    width: '40px',
    height: '20px',
    border: '1px solid #dcdfe6',
    outline: 'none',
    borderRadius: '10px',
    background: '#dcdfe6',
    cursor: 'pointer',
    transition: 'all 0.3s ease-in-out',
    verticalAlign: 'middle',
    '&:after': {
      content: '""',
      position: 'absolute',
      top: '2px',
      left: '1px',
      borderRadius: '100%',
      transition: 'all 0.3s ease-in-out',
      width: '16px',
      height: '16px',
      backgroundColor: '#fff',
    },
  },
  'switch--active': {
    borderColor: '#409eff',
    backgroundColor: '#409eff',
    '&:after': {
      left: '100%',
      marginLeft: '-17px',
    },
  },
  'switch--disabled': {
    cursor: 'not-allowed',
    opacity: 0.5,
  },
};
