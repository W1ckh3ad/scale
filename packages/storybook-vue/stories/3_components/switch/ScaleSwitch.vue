<template>
  <scale-switch
    :custom-class="customClass"
    :disabled="disabled"
    :input-id="inputId"
    :checked="checked"
    @scaleChange="scaleChange"
  >
  </scale-switch>
</template>

<script>
import { action } from "@storybook/addon-actions";

export default {
  props: {
    customClass: String,
    disabled: { type: Boolean, default: false },
    inputId: { type: String },
    checked: { type: Boolean, default: false }
  },
  methods: {
    scaleChange($event) {
      action("scaleChange");
      this.$emit("scaleChange", $event);
    }
  }
};
</script>
