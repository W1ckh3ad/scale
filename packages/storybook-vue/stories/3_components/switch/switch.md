# Switch

With switches, users can easily toggle between the on and off state of a setting.

## General

### When to Use

Switches are only used for binary actions that occur immediately after the user “flips” the switch. They are commonly used for “on/off” switches.

![Image Name](/assets/3_components/switch/image-20200810093225678.png)

## Elemets

### Label

Clearly label the switch so users know what setting they are manipulating e.g. "dark mode".

## Related Components

Checkbox, Select Box, Radio Button

## Code

