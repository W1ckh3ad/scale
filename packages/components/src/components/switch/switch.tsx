import {
  Component,
  h,
  Prop,
  Host,
  Event,
  EventEmitter,
  Listen,
} from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { styles } from './switch.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

let i = 0;

@Component({
  tag: 'scale-switch',
  shadow: false,
})
export class Switch implements Base {
  /** (optional) Switch class */
  @Prop() customClass?: string = '';
  /** (optional) Active switch */
  @Prop({ reflect: true }) checked?: boolean = false;
  /** (optional) Disabled switch */
  @Prop() disabled?: boolean = false;
  /** (optional) Input id */
  @Prop() inputId?: string;
  /** (optional) switch label */
  @Prop() label?: string;
  /** (optional) Injected jss styles */
  @Prop() styles?: any;
  /** decorator Jss stylesheet */
  @CssInJs('Switch', styles) stylesheet: StyleSheet;

  /** Emitted when the switch was clicked */
  @Event() scaleChange!: EventEmitter;

  @Listen('click', { capture: true })
  handleClick() {
    if (!this.disabled) {
      this.checked = !this.checked;
    }
  }

  componentWillLoad() {
    if (this.inputId == null) {
      this.inputId = 'switch-' + i++;
    }
  }
  componentWillUpdate() {}
  componentDidUnload() {}

  render() {
    const { classes } = this.stylesheet;
    return (
      <Host>
        <div class={this.getCssClassMap()}>
          <input
            type="checkbox"
            checked={this.checked}
            disabled={this.disabled}
            aria-pressed={this.checked}
            aria-labelledby={`${this.inputId}-label`}
            id={this.inputId}
            onChange={() => {
              if (!this.disabled) {
                this.scaleChange.emit({ value: this.checked });
              }
            }}
          />
          <div
            class={classes['switch__wrapper']}
            onClick={() => {
              if (!this.disabled) {
                this.scaleChange.emit({ value: this.checked });
              }
            }}
          >
            <div class={classes['switch__container']}>
              <div class={classes['switch__toggle']} />
              <div class={classes['switch__text']} />
            </div>
            {!!this.label && (
              <label
                class={classes['switch__label']}
                id={`${this.inputId}-label`}
                htmlFor={this.inputId}
                onClick={event => {
                  event.preventDefault();
                }}
              >
                {this.label}
              </label>
            )}
          </div>
        </div>
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(classes.switch, this.customClass && this.customClass);
  }
}
