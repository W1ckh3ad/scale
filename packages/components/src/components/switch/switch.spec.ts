import { newSpecPage } from '@stencil/core/testing';
import { Switch } from './switch';
import { styles } from './switch.styles';
import jss from 'jss';

describe('Switch', () => {
  let element;
  let stylesheet;
  beforeEach(async () => {
    element = new Switch();
    stylesheet = element.stylesheet = jss.createStyleSheet(styles as any);
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Switch],
      html: `<scale-switch></scale-switch>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should have toggle active state', () => {
    element.disabled = false;
    element.active = true;
    element.toggleSwitch();
    expect(element.active).toBe(false);
  });

  it('should handle css classes', () => {
    element.customClass = 'custom';
    expect(element.getCssClassMap()).toContain('custom');

    element.active = true;
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['switch--active']
    );

    element.disabled = true;
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['switch--disabled']
    );
  });
});
