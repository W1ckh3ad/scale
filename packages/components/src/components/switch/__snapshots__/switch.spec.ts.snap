// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should match snapshot 1`] = `
<scale-switch>
  <mock:shadow-root>
    <style>
      .switch-0-0-4 {
  width: 40px;
  border: 1px solid #dcdfe6;
  cursor: pointer;
  height: 20px;
  margin: 0;
  display: inline-block;
  outline: none;
  position: relative;
  background: #dcdfe6;
  transition: all 0.3s ease-in-out;
  border-radius: 10px;
  vertical-align: middle;
}
.switch-0-0-4:after {
  top: 2px;
  left: 1px;
  width: 16px;
  height: 16px;
  content: "";
  position: absolute;
  transition: all 0.3s ease-in-out;
  border-radius: 100%;
  background-color: #fff;
}
.switch--active-0-0-5 {
  border-color: #409eff;
  background-color: #409eff;
}
.switch--active-0-0-5:after {
  left: 100%;
  margin-left: -17px;
}
.switch--disabled-0-0-6 {
  cursor: not-allowed;
  opacity: 0.5;
}
    </style>
    <div class="switch-0-0-4"></div>
  </mock:shadow-root>
</scale-switch>
`;
